sdk/cosmosdb/cosmos/src/routing/partitionKeyRangeCache.ts

```diff
@@ -68,4 +72,35 @@ export class PartitionKeyRangeCache {
     );
     return createCompleteRoutingMap(resources.map((r) => [r, true]));
   }
+
+  /**
+   * Given a partition key, returns the partition key range id
+   * @internal
+   */
+  public async getPartitionKeyRangeIdFromPartitionKey(
+    collectionLink: string,
+    partitionKey: PartitionKeyInternal,
+    partitionKeyDefinition: PartitionKeyDefinition,
+    diagnosticNode: DiagnosticNodeInternal,
+  ): Promise<string | undefined> {
+    let partitionKeyRangeId: string;
+    const hashedPartitionKey = hashPartitionKey(partitionKey, partitionKeyDefinition);
+    const partitionKeyRanges = (
+      await this.onCollectionRoutingMap(collectionLink, diagnosticNode)
+    ).getOrderedParitionKeyRanges();
+
+    for (const partitionKeyRange of partitionKeyRanges) {
```

Comment:
We can do better here since we do know the min and Max ranges. Maybe we can use some other data structure like Binary search tree to search individual elements.